extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchMergeBufferAggregator.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation.datasketches.hll;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+
+import com.google.common.util.concurrent.Striped;
+import com.yahoo.memory.WritableMemory;
+import com.yahoo.sketches.hll.HllSketch;
+import com.yahoo.sketches.hll.TgtHllType;
+import com.yahoo.sketches.hll.Union;
+
+import org.apache.druid.query.aggregation.BufferAggregator;
+import org.apache.druid.segment.ColumnValueSelector;
+
+/**
+ * This aggregator merges existing sketches.
+ * The input column must contain {@link HllSketch}
+ * @author Alexander Saydakov
+ */
+public class HllSketchMergeBufferAggregator implements BufferAggregator
+{
+
+  private static final int NUM_STRIPES = 64; // for locking per buffer position (power of 2 to make index computation faster)
+
+  private final ColumnValueSelector<HllSketch> selector;
+  private final int lgK;
+  private final TgtHllType tgtHllType;
+  private final int size;
+  private final Striped<ReadWriteLock> stripedLock = Striped.readWriteLock(NUM_STRIPES);
+
+  public HllSketchMergeBufferAggregator(
+      final ColumnValueSelector<HllSketch> selector,
+      final int lgK,
+      final TgtHllType tgtHllType,
+      final int size
+  )
+  {
+    this.selector = selector;
+    this.lgK = lgK;
+    this.tgtHllType = tgtHllType;
+    this.size = size;
+  }
+
+  @Override
+  public void init(final ByteBuffer buf, final int position)
+  {
+    final WritableMemory mem = WritableMemory.wrap(buf).writableRegion(position, size);
+    new Union(lgK, mem);
```

Comment:
In a sense. We are given a slot in a buffer, and we need to initialize an aggregation object there. Our sketches have a special mode for this (so called "direct" as opposed to "heap"). In this case the Union constructor is given a chunk of memory to occupy. It will initialize a few things there so that later this chunk can be "wrapped" (quickly interpreted as a Union object) and updated:
Union union = Union.writableWrap(mem);
union.update(sketch);
In this mode it is not necessary to keep the constructed object around because it is cheap to reconstruct from the chunk of memory using the "wrapping" operation. Perhaps I should add a comment in the code to explain this.

Reply:
Yes please add a short comment that this constructor changes the underlying memory, and not storing it in any variable is not a mistake.